File: 59_spiral_matrix_II/spiral_matrix_II.py
class Solution(object):
    def generateMatrix(self, n):
        """
        : type n: int
        : rtype: List[List[int]]
        """
        array = [i + 1 for i in range(0, n**2)]

        result = [[0 for i in range(0, n)] for j in range(0, n)]

        beginX = 0
        endX = n - 1

        beginY = 0
        endY = n - 1

        count = 0


        while True:
            # From left to right
            for i in range(beginX, endX + 1):
            	result[beginY][i] = array[count]
            	count += 1
            beginY += 1
            if beginY > endY:
            	break
            # From top to bottom
            for j in range(beginY, endY + 1):
            	result[j][endX] = array[count]
            	count += 1
            endX -= 1
            if endX < beginX:
            	break
            # From right to left
            for i in reversed(range(beginX, endX + 1)):
            	result[endY][i] = array[count]
            	count += 1
            endY -= 1
            if endY < beginY:
            	break
            # From bottom to top
            for j in reversed(range(beginY, endY + 1)):
            	result[j][beginX] = array[count]
            	count += 1
            beginX += 1 
            if beginX > endX:
            	break

        return result


if __name__ == '__main__':
	sol = Solution()
	n = 6;
	result = sol.generateMatrix(n)
	print result

File: 59_spiral_matrix_II/spiral_matrix_II.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int> > res(n, vector<int>(n, 0));
        if(n == 0) { return res; }

        int beginX = 0, endX = n - 1;
        int beginY = 0, endY = n - 1;
        int count = 1;

        while(true) {
            for(int i = beginX; i <= endX; i++) {
                res[beginY][i] = count;
                count++;
            }

            if(++beginY > endY) { break; }

            for(int i = beginY; i <= endY; i++) {
                res[i][endX] = count;
                count++;
            }

            if(--endX < beginX) { break; }

            for(int i = endX; i >= beginX; i--) {
                res[endY][i] = count;
                count++;
            }

            if(--endY < beginY) { break; }

            for(int i = endY; i >= beginY; i--) {
                res[i][beginX] = count;
                count++;
            }

            if(++beginX > endX) { break; }
        }

        return res;
    }
};

int main()
{
	int  n = 3;
	Solution sol;
	vector<vector<int> > result;
	result = sol.generateMatrix(n);

	for(vector<vector<int> >::iterator it = result.begin();
		it != result.end();
		it++)
	{
		for(vector<int>::iterator itt = (*it).begin();
			itt != (*it).end();
			itt++)
		{
			cout << *(itt) << " ";
		}
		cout << endl;
	}

}